### PR TITLE
Jvm Spawn

### PR DESCRIPTION
**Pull request by emilbaekdahl**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 1]
```diff
@@ -704,6 +705,53 @@ object GenExpression {
...
+
+      //#3  +3
+      visitor.visitTypeInsn(NEW, JvmName.Thread.toInternalName)
+
+      visitor.visitInsn(DUP)
+
+      visitor.visitVarInsn(ALOAD, 1)
+
+      //#4  +4
+      visitor.visitMethodInsn(INVOKESPECIAL, JvmName.Thread.toInternalName, "<init>", AsmOps.getMethodDescriptor(List(JvmType.Runnable), JvmType.Void), false)
```
Comment (jespersandnielsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Add name for thread

File: main/src/ca/uwaterloo/flix/language/phase/jvm/RunnableSpawn.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+    public void run() {
+        try{
+            fn.call();
+        } catch (java.lang.Exception ex) {
+
+        }
+    }
+}
+
```
Comment (jespersandnielsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/RunnableSpawn.java]: Remove this file

File: main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala [hunk 3]
```diff
@@ -20,6 +20,7 @@ import scala.collection.JavaConverters._
 import java.lang.reflect.Modifier
 import java.util.concurrent.locks.{Condition, Lock, ReentrantLock}
 import java.util
+import scala.collection.mutable.ListBuffer
```
Comment (jespersandnielsen) [inline main/src/ca/uwaterloo/flix/runtime/interpreter/Interpreter.scala]: Remove this import - we don't use it anymore

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala [hunk 4]
```diff
@@ -704,6 +705,53 @@ object GenExpression {
...
+      compileExpression(exp, visitor, currentClass, lenv0, entryPoint)
+
+      //visitor.visitInsn(DUP)
+
+      // Invoking the constructor
+      //#2  +3
+      visitor.visitMethodInsn(INVOKESPECIAL, JvmName.RunnableSpawn.toInternalName, "<init>", AsmOps.getMethodDescriptor(List(functionType), JvmType.Void), false)
+
+
+      visitor.visitVarInsn(ASTORE, 1)
```
Comment (jespersandnielsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala]: Dont use `ASTORE` use the stack instead

**Review by jespersandnielsen — DISMISSED** [review]

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenClosureClasses.scala [hunk 5]
```diff
@@ -110,12 +110,60 @@ object GenClosureClasses {
...
+
+      copyMethod.visitVarInsn(ALOAD, 0)
+      copyMethod.visitFieldInsn(GETFIELD, classType.name.toInternalName, s"arg$ind", erasedType.toDescriptor)
+
+      copyMethod.visitMethodInsn(INVOKEVIRTUAL, classType.name.toInternalName, s"setArg$ind",
+        AsmOps.getMethodDescriptor(List(erasedType), JvmType.Void), false)
+    }
+
+    copyMethod.visitInsn(ARETURN)
+
```
Comment (jacobtonder) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenClosureClasses.scala]: Remove newline

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    AsmOps.compileField(visitor, "ctx", JvmType.Context, isStatic = false, isPrivate = true)
+
+    // Instance field
+    AsmOps.compileField(visitor, "fn", functionType, isStatic = false, isPrivate = true)
+
+    // Generate the constructor
+    genConstructor(visitor)
+
+    // Generate `toString` method
+    //genToString(visitor)
```
Comment (jacobtonder) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala]: Remove

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+    // Generate `equals` method
+    //genEquals(visitor)
```
Comment (jacobtonder) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala]: Remove

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    initMethod.visitVarInsn(ALOAD, 1)
+    initMethod.visitVarInsn(ALOAD, 0)
+    initMethod.visitFieldInsn(GETFIELD, JvmName.RunnableSpawn.toInternalName, "ctx", JvmType.Context.toDescriptor)
+    initMethod.visitMethodInsn(INVOKEINTERFACE, functionType.name.toInternalName, "copy",
+      AsmOps.getMethodDescriptor(List(JvmType.Context), functionType), true)
+    initMethod.visitFieldInsn(PUTFIELD, JvmName.RunnableSpawn.toInternalName, "fn", functionType.toDescriptor)
+    initMethod.visitInsn(RETURN)
+    initMethod.visitMaxs(2, 2)
+    initMethod.visitEnd()
+  }
```
Comment (jacobtonder) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala]: new line

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    initMethod.visitMethodInsn(INVOKESPECIAL, JvmName.Context.toInternalName, "<init>", AsmOps.getMethodDescriptor(Nil, JvmType.Void), false)
+    initMethod.visitFieldInsn(PUTFIELD, JvmName.RunnableSpawn.toInternalName, "ctx", JvmType.Context.toDescriptor)
+
+    initMethod.visitVarInsn(ALOAD, 0)
+    initMethod.visitVarInsn(ALOAD, 1)
+    initMethod.visitVarInsn(ALOAD, 0)
+    initMethod.visitFieldInsn(GETFIELD, JvmName.RunnableSpawn.toInternalName, "ctx", JvmType.Context.toDescriptor)
+    initMethod.visitMethodInsn(INVOKEINTERFACE, functionType.name.toInternalName, "copy",
+      AsmOps.getMethodDescriptor(List(JvmType.Context), functionType), true)
+    initMethod.visitFieldInsn(PUTFIELD, JvmName.RunnableSpawn.toInternalName, "fn", functionType.toDescriptor)
```
Comment (jacobtonder) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala]: New line

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    val functionType = JvmOps.getFunctionInterfaceType(Type.mkArrow(Type.Unit, Type.Unit))
+    val runMethod = visitor.visitMethod(ACC_PUBLIC, "run", AsmOps.getMethodDescriptor(Nil, JvmType.Void), null, null)
+    runMethod.visitCode()
+    runMethod.visitVarInsn(ALOAD, 0)
+    runMethod.visitFieldInsn(GETFIELD, JvmName.RunnableSpawn.toInternalName, "fn", functionType.toDescriptor)
+
+    runMethod.visitVarInsn(ALOAD, 0)
+    runMethod.visitFieldInsn(GETFIELD, JvmName.RunnableSpawn.toInternalName, "ctx", JvmType.Context.toDescriptor)
+    runMethod.visitMethodInsn(INVOKEINTERFACE, functionType.name.toInternalName, "apply",
+      AsmOps.getMethodDescriptor(List(JvmType.Context), JvmType.Void), true)
```
Comment (jacobtonder) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala]: new line

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    // Set context
+    initMethod.visitVarInsn(ALOAD, 0)
+    initMethod.visitTypeInsn(NEW, JvmName.Context.toInternalName)
+    initMethod.visitInsn(DUP)
+    initMethod.visitMethodInsn(INVOKESPECIAL, JvmName.Context.toInternalName, "<init>", AsmOps.getMethodDescriptor(Nil, JvmType.Void), false)
+    initMethod.visitFieldInsn(PUTFIELD, JvmName.RunnableSpawn.toInternalName, "ctx", JvmType.Context.toDescriptor)
+
+    initMethod.visitVarInsn(ALOAD, 0)
+    initMethod.visitVarInsn(ALOAD, 1)
+    initMethod.visitVarInsn(ALOAD, 0)
```
Comment (jacobtonder) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenRunnableSpawn.scala]: Load these when needed, instead of loading them all at the same time.

**Review by jacobtonder — CHANGES_REQUESTED** [review]

**Review by jacobtonder — APPROVED** [review]

**Pull request merged**